src/platform_impl/linux/common/xkb_state.rs

```diff
@@ -0,0 +1,774 @@
+            // init of compose table failed, continue without compose
+            return;
+        }
+
+        let compose_state = (XKBCH.xkb_compose_state_new)(
+            compose_table,
+            ffi::xkb_compose_state_flags::XKB_COMPOSE_STATE_NO_FLAGS,
+        );
+
+        if compose_table.is_null() {
+            // init of compose state failed, continue without compose
+            (XKBCH.xkb_compose_table_unref)(compose_table);
+            return;
+        }
+
+        let compose_state_2 = (XKBCH.xkb_compose_state_new)(
+            compose_table,
+            ffi::xkb_compose_state_flags::XKB_COMPOSE_STATE_NO_FLAGS,
+        );
+
+        if compose_state_2.is_null() {
+            // init of compose state failed, continue without compose
+            (XKBCH.xkb_compose_table_unref)(compose_table);
+            (XKBCH.xkb_compose_state_unref)(compose_state);
+            return;
+        }
+
+        self.xkb_compose_table = compose_table;
+        self.xkb_compose_state = compose_state;
+        self.xkb_compose_state_2 = compose_state_2;
+    }
+
+    unsafe fn post_init(&mut self, state: *mut ffi::xkb_state, keymap: *mut ffi::xkb_keymap) {
+        self.xkb_keymap = keymap;
+        self.xkb_state = state;
+        self.mods_state.update_with(state);
+    }
+
+    unsafe fn de_init(&mut self) {
+        (XKBH.xkb_state_unref)(self.xkb_state);
+        self.xkb_state = ptr::null_mut();
+        (XKBH.xkb_keymap_unref)(self.xkb_keymap);
+        self.xkb_keymap = ptr::null_mut();
+    }
+
+    #[cfg(feature = "x11")]
+    pub(crate) unsafe fn init_with_x11_keymap(&mut self) {
+        if !self.xkb_keymap.is_null() {
+            self.de_init();
+        }
+
+        // TODO: Support keyboards other than the "virtual core keyboard device".
+        let core_keyboard_id = (XKBXH.xkb_x11_get_core_keyboard_device_id)(self.xcb_connection);
+        let keymap = (XKBXH.xkb_x11_keymap_new_from_device)(
+            self.xkb_context,
+            self.xcb_connection,
+            core_keyboard_id,
+            xkbcommon_dl::xkb_keymap_compile_flags::XKB_KEYMAP_COMPILE_NO_FLAGS,
+        );
+        assert_ne!(keymap, ptr::null_mut());
+
+        let state =
+            (XKBXH.xkb_x11_state_new_from_device)(keymap, self.xcb_connection, core_keyboard_id);
+        self.post_init(state, keymap);
+    }
+
+    #[cfg(feature = "wayland")]
+    pub(crate) unsafe fn init_with_fd(&mut self, fd: File, size: usize) {
+        if !self.xkb_keymap.is_null() {
+            self.de_init();
+        }
+
+        let map = MmapOptions::new().len(size).map(&fd).unwrap();
+
+        let keymap = (XKBH.xkb_keymap_new_from_string)(
+            self.xkb_context,
+            map.as_ptr() as *const _,
+            ffi::xkb_keymap_format::XKB_KEYMAP_FORMAT_TEXT_V1,
+            ffi::xkb_keymap_compile_flags::XKB_KEYMAP_COMPILE_NO_FLAGS,
+        );
+
+        if keymap.is_null() {
+            panic!("Received invalid keymap from compositor.");
```

Comment:
I chose to panic here because it was less work than bubbling the error up to the user somehow, although it's not obvious to me how we'd do that and still let the user use their window(s) in this degraded state. Another option would've been to log the error and move on with keyboard events remaining non-functional, but that makes the error one you have to dig through logs to find. I also don't think this will come up often in practice, but that's just a guess.